### PR TITLE
automated tests and scope program

### PR DESCRIPTION
**Pull request by mjaskula**
This PR sets us up to be able to run automated unit tests locally in CPython (not micropython), and includes example tests for both a script and the EuroPi lib.

This PR also includes:
- an oscilloscope program, a portion of which has unit tests
- docs for running tests and installing test requirements
- updates to docs for installing the docs requirements

This PR does not include running the tests in a CI environment on commit, this will be done under a separate PR.

### CODE REVIEW

File: software/README.md [hunk 1]
```diff
@@ -1,3 +1,33 @@
 # EuroPi Software
 
-See the EuroPi firmware API documentation here: https://allen-synthesis.github.io/EuroPi/generated/europi.html
\ No newline at end of file
+See the EuroPi firmware API documentation here: https://allen-synthesis.github.io/EuroPi/generated/europi.html
+
+
+
+
```
Comment (awonak) [inline software/README.md:7]: nit: remove extraneous newlines.

```suggestion

```

File: software/firmware/contrib/scope.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
+from time import sleep
+
+try:
+    # local dev
+    from software.firmware.europi import ain, b1, b2, cv1, cv2, cv4, din, k1, k2, oled
+    from software.firmware.europi import reset_state, MAX_INPUT_VOLTAGE, OLED_HEIGHT, OLED_WIDTH
```
Comment (awonak) [inline software/firmware/contrib/scope.py:6]: remove unused function

```suggestion
    from software.firmware.europi import MAX_INPUT_VOLTAGE, OLED_HEIGHT, OLED_WIDTH
```

File: software/tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import sys
+from pathlib import Path
+
+sys.path.append(str(Path(__file__).parent.parent / 'firmware'))
```
Comment (awonak) [inline software/tests/conftest.py:4]: Can you help me understand this syntax? How is there a `/` in the middle of `str(...)`?

**Review by awonak — APPROVED** [review]
Excellent work here, I'm excited to be able to start writing tests against the firmware!

File: software/tests/conftest.py (hunk 3 above)
Comment (mjaskula) [reply software/tests/conftest.py:4]: The left operand isn't a string, it's a `Path`, and that class overrides the `/` operator to allow this syntax. The resulting `Path` object is then converted into a string.

I had to look it up too when I found the example code that I copied this from.

Comment (awonak) [reply software/tests/conftest.py:4]: Lol, python being python.

**Pull request merged**